### PR TITLE
fix: Add support for unrecoverable exception

### LINKED ISSUE #34: Allow to stop retrying an SQS message
Hey there,

By browsing the code source around the `SqsConsumer` and the `SqsTransport` I could not find a way to "cancel" an event (delete it or send it to a dead letter queue). 

Symfony messenger allow [to throw an `UnrecoverableMessageHandlingException` exception to stop](https://symfony.com/doc/current/messenger.html#avoiding-retrying) to stop retrying a message. It could be nice to have this behaviour implemented in the bref SQS bridge. 

What did you think?

### PR DESCRIPTION
**Pull request by tyx**
Fix #34 

I just realized this one was never fixed and face to this issue today so here we go !

### CODE REVIEW

**Review by shouze — APPROVED** [review]
:heart_eyes:

**Comment by mnapoli** [comment]
LGTM, just want to validate with you all: that means that the error will not be reported in the Lambda execution status nor in the SQS job (it will not go in a DLQ).

The only way to catch this is to explicitly look for a specific string in logs.

That's the intended behavior, right?

File: src/Service/Sqs/SqsConsumer.php [hunk 1]
```diff
@@ -89,6 +90,9 @@ public function handleSqs(SqsEvent $event, Context $context): void
                     $stamps[] = new AmazonSqsXrayTraceHeaderStamp($context->getTraceId());
                 }
                 $this->busDriver->putEnvelopeOnBus($this->bus, $envelope->with(...$stamps), $this->transportName);
+            } catch (UnrecoverableExceptionInterface $exception) {
+                $this->logger->error(sprintf('SQS record with id "%s" failed to be processed. But failure was marked as unrecoverable. Message will be acknowledged.', $record->getMessageId()));
+                $this->logger->error($exception->getMessage());
```
Comment (starred-gijs) [inline src/Service/Sqs/SqsConsumer.php:95]: I would like to suggest to match the [messenger error level/message code](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Messenger/EventListener/SendFailedMessageForRetryListener.php#L82) a little:
```
$this->logger?->critical('Error thrown while handling message {class}. Removing from transport after {retryCount} retries. Error: "{error}"', $context + ['retryCount' => $retryCount, 'error' => $throwable->getMessage(), 'exception' => $throwable]);
```

```suggestion
                $this->logger->critical('Error thrown while handling message {class}. Removing from transport. Error: "{error}"', ['error' => $throwable->getMessage(), 'exception' => $exception]);
```

**Comment by starred-gijs** [comment]
> it will not go in a DLQ
Im now a bit in doubt if this is the same behaviour with the Messenger worker, I will do some tests tomorrow :)

**Comment by tyx** [comment]
> LGTM, just want to validate with you all: that means that the error will not be reported in the Lambda execution status nor in the SQS job (it will not go in a DLQ).
> 
> The only way to catch this is to explicitly look for a specific string in logs.
> 
> That's the intended behavior, right?

Exactly. This is the Messenger behavior through this subscriber : 
https://github.com/symfony/messenger/blob/c1609b6689b07890e94cf3016396f550c5feeee9/EventListener/SendFailedMessageForRetryListener.php

When a message fail through the consume command, Messenger ask if it need to be retried. And failure of kind Unrecoverable return false to this question.

File: src/Service/Sqs/SqsConsumer.php (hunk 1 above)
Comment (tyx) [reply src/Service/Sqs/SqsConsumer.php:95]: I mostly be consistent with log in the other catch statement. But I'm OK to improve both of them ;)

Comment (starred-gijs) [reply src/Service/Sqs/SqsConsumer.php:95]: I see now I also made a mistake, the `{class}` is not yet set in context, please ignore my suggestion. I would like to see the exception in the `exception` [context](https://www.php-fig.org/psr/psr-3/#13-context)

**Comment by starred-gijs** [comment]
> Just to complete, Symfony has a "failure" transport system and yes Unrecoverable exception can go to it.

Yes that is what I was expecting. I actually get this behaviour with https://github.com/Happyr/bref-messenger-failure-strategies.

**Review by starred-gijs — APPROVED** [review]

**Comment by mnapoli** [comment]
Thanks for the review everyone, merging now!

**Pull request merged**